File: sdk/csharp-holons/Holons/Identity.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace Holons;

public static class Identity
{
    public const string ProtoManifestFileName = "holon.proto";

    public static IdentityParser.HolonIdentity ParseHolon(string path)
    {
        return IdentityParser.ParseHolon(path);
    }

    public static IdentityParser.ResolvedManifest Resolve(string root)
    {
        return ResolveProtoFile(ResolveManifestPath(root));
    }

    public static IdentityParser.ResolvedManifest ResolveProtoFile(string path)
    {
        var resolvedPath = Path.GetFullPath(path);
        if (!string.Equals(Path.GetFileName(resolvedPath), ProtoManifestFileName, StringComparison.Ordinal))
            throw new IOException($"{resolvedPath} is not a {ProtoManifestFileName} file");

        return IdentityParser.ParseManifest(resolvedPath) with { SourcePath = resolvedPath };
    }

    public static string? FindHolonProto(string root)
    {
        return IdentityParser.FindHolonProto(root);
    }

    public static string ResolveManifestPath(string root)
    {
        return IdentityParser.ResolveManifestPath(root);
    }
}

/// <summary>Parse holon.proto identity files.</summary>
public static class IdentityParser
{
    private static readonly Regex ManifestPattern = new(
        @"option\s*\(\s*holons\.v1\.manifest\s*\)\s*=\s*\{",
        RegexOptions.Compiled);

    /// <summary>Parsed holon identity.</summary>
    public record HolonIdentity
    {
        public string Uuid { get; init; } = "";
        public string GivenName { get; init; } = "";
        public string FamilyName { get; init; } = "";
        public string Motto { get; init; } = "";
        public string Composer { get; init; } = "";
        public string Clade { get; init; } = "";
        public string Status { get; init; } = "";
        public string Born { get; init; } = "";
        public string Lang { get; init; } = "";
        public List<string> Parents { get; init; } = new();
        public string Reproduction { get; init; } = "";
        public string GeneratedBy { get; init; } = "";
        public string ProtoStatus { get; init; } = "";
        public List<string> Aliases { get; init; } = new();

        public string Slug()
        {
            var given = (GivenName ?? "").Trim();
            var family = (FamilyName ?? "").Trim().TrimEnd('?');
            if (given.Length == 0 && family.Length == 0)
                return "";

            return $"{given}-{family}"
                .Trim()
                .ToLowerInvariant()
                .Replace(" ", "-", StringComparison.Ordinal)
                .Trim('-');
        }
    }

    public record ResolvedManifest
    {
        public HolonIdentity Identity { get; init; } = new();
        public string SourcePath { get; init; } = "";
        public string Kind { get; init; } = "";
        public string BuildRunner { get; init; } = "";
        public string BuildMain { get; init; } = "";
        public string ArtifactBinary { get; init; } = "";
        public string ArtifactPrimary { get; init; } = "";
    }

    /// <summary>Parse a holon.proto file.</summary>
    public static HolonIdentity ParseHolon(string path)
    {
        return ParseManifest(path).Identity;
    }

    public static ResolvedManifest ParseManifest(string path)
    {
        var text = File.ReadAllText(path);
        var manifestBlock = ExtractManifestBlock(text)
            ?? throw new FormatException($"{path}: missing holons.v1.manifest option in holon.proto");

        var identityBlock = ExtractBlock("identity", manifestBlock) ?? "";
        var lineageBlock = ExtractBlock("lineage", manifestBlock) ?? "";
        var buildBlock = ExtractBlock("build", manifestBlock) ?? "";
        var artifactsBlock = ExtractBlock("artifacts", manifestBlock) ?? "";

        return new ResolvedManifest
        {
            Identity = new HolonIdentity
            {
                Uuid = Scalar("uuid", identityBlock),
                GivenName = Scalar("given_name", identityBlock),
                FamilyName = Scalar("family_name", identityBlock),
                Motto = Scalar("motto", identityBlock),
                Composer = Scalar("composer", identityBlock),
                Clade = Scalar("clade", identityBlock),
                Status = Scalar("status", identityBlock),
                Born = Scalar("born", identityBlock),
                Lang = Scalar("lang", manifestBlock),
                Parents = StringList("parents", lineageBlock),
                Reproduction = Scalar("reproduction", lineageBlock),
                GeneratedBy = Scalar("generated_by", lineageBlock),
                ProtoStatus = Scalar("proto_status", identityBlock),
                Aliases = StringList("aliases", identityBlock),
            },
            Kind = Scalar("kind", manifestBlock),
            BuildRunner = Scalar("runner", buildBlock),
            BuildMain = Scalar("main", buildBlock),
            ArtifactBinary = Scalar("binary", artifactsBlock),
            ArtifactPrimary = Scalar("primary", artifactsBlock),
            SourcePath = Path.GetFullPath(path),
        };
    }

    public static string? FindHolonProto(string root)
    {
        var resolved = Path.GetFullPath(root);
        if (File.Exists(resolved))
            return string.Equals(Path.GetFileName(resolved), "holon.proto", StringComparison.Ordinal) ? resolved : null;
        if (!Directory.Exists(resolved))
            return null;

        var direct = Path.Combine(resolved, "holon.proto");
        if (File.Exists(direct))
            return direct;

        var apiV1 = Path.Combine(resolved, "api", "v1", "holon.proto");
        if (File.Exists(apiV1))
            return apiV1;

        return Directory.EnumerateFiles(resolved, "holon.proto", SearchOption.AllDirectories)
            .OrderBy(path => path, StringComparer.Ordinal)
            .FirstOrDefault();
    }

    public static string ResolveManifestPath(string root)
    {
        var resolved = Path.GetFullPath(root);
        var searchRoots = new List<string> { resolved };
        var parent = Path.GetDirectoryName(resolved);
        if (string.Equals(Path.GetFileName(resolved), "protos", StringComparison.Ordinal) && !string.IsNullOrWhiteSpace(parent))
        {
            searchRoots.Add(parent);
        }
        else if (!string.IsNullOrWhiteSpace(parent) && !string.Equals(parent, resolved, StringComparison.Ordinal))
        {
            searchRoots.Add(parent);
        }

        foreach (var searchRoot in searchRoots)
        {
            var candidate = FindHolonProto(searchRoot);
            if (!string.IsNullOrWhiteSpace(candidate))
                return candidate;
        }

        throw new IOException($"no holon.proto found near {resolved}");
    }

    private static string? ExtractManifestBlock(string source)
    {
        var match = ManifestPattern.Match(source);
        if (!match.Success)
            return null;
        var braceIndex = source.IndexOf('{', match.Index);
        return braceIndex >= 0 ? BalancedBlockContents(source, braceIndex) : null;
    }

    private static string? ExtractBlock(string name, string source)
    {
        var match = Regex.Match(source, $@"\b{Regex.Escape(name)}\s*:\s*\{{");
        if (!match.Success)
            return null;
        var braceIndex = source.IndexOf('{', match.Index);
        return braceIndex >= 0 ? BalancedBlockContents(source, braceIndex) : null;
    }

    private static string Scalar(string name, string source)
    {
        var quoted = Regex.Match(source, $@"\b{Regex.Escape(name)}\s*:\s*""((?:[^""\\]|\\.)*)""");
        if (quoted.Success)
            return UnescapeProtoString(quoted.Groups[1].Value);

        var bare = Regex.Match(source, $@"\b{Regex.Escape(name)}\s*:\s*([^\s,\]\}}]+)");
        return bare.Success ? bare.Groups[1].Value : "";
    }

    private static List<string> StringList(string name, string source)
    {
        var body = Regex.Match(source, $@"(?s)\b{Regex.Escape(name)}\s*:\s*\[(.*?)\]");
        if (!body.Success)
            return new List<string>();

        return Regex.Matches(body.Groups[1].Value, @"""((?:[^""\\]|\\.)*)""|([^\s,\]]+)")
            .Select(match => match.Groups[1].Success
                ? UnescapeProtoString(match.Groups[1].Value)
                : match.Groups[2].Value)
            .Where(value => !string.IsNullOrEmpty(value))
            .ToList();
    }

    private static string? BalancedBlockContents(string source, int openingBrace)
    {
        var depth = 0;
        var insideString = false;
        var escaped = false;
        var contentStart = openingBrace + 1;

        for (var index = openingBrace; index < source.Length; index++)
        {
            var ch = source[index];
            if (insideString)
            {
                if (escaped)
                {
                    escaped = false;
                }
                else if (ch == '\\')
                {
                    escaped = true;
                }
                else if (ch == '"')
                {
                    insideString = false;
                }
                continue;
            }

            if (ch == '"')
            {
                insideString = true;
            }
            else if (ch == '{')
            {
                depth += 1;
            }
            else if (ch == '}')
            {
                depth -= 1;
                if (depth == 0)
                    return source[contentStart..index];
            }
        }

        return null;
    }

    private static string UnescapeProtoString(string value)
    {
        return value.Replace("\\\"", "\"", StringComparison.Ordinal)
            .Replace("\\\\", "\\", StringComparison.Ordinal);
    }
}

File: sdk/python-holons/holons/identity.py
from __future__ import annotations

"""Resolve holon manifest data from holon.proto files."""

from dataclasses import dataclass, field
from pathlib import Path
import re

PROTO_MANIFEST_FILE_NAME = "holon.proto"

_MANIFEST_OPTION_RE = re.compile(
    r"option\s*\(\s*holons\.v1\.manifest\s*\)\s*=\s*\{",
    re.DOTALL,
)


@dataclass
class HolonIdentity:
    """Parsed identity from a holon.proto manifest."""

    uuid: str = ""
    given_name: str = ""
    family_name: str = ""
    motto: str = ""
    composer: str = ""
    clade: str = ""
    status: str = ""
    born: str = ""
    version: str = ""
    lang: str = ""
    parents: list[str] = field(default_factory=list)
    reproduction: str = ""
    generated_by: str = ""
    proto_status: str = ""
    aliases: list[str] = field(default_factory=list)

    def slug(self) -> str:
        given = self.given_name.strip()
        family = self.family_name.strip().removesuffix("?")
        if not given and not family:
            return ""
        return f"{given}-{family}".strip().lower().replace(" ", "-").strip("-")


@dataclass
class ResolvedSkill:
    name: str = ""
    description: str = ""
    when: str = ""
    steps: list[str] = field(default_factory=list)


@dataclass
class ResolvedSequenceParam:
    name: str = ""
    description: str = ""
    required: bool = False
    default: str = ""


@dataclass
class ResolvedSequence:
    name: str = ""
    description: str = ""
    params: list[ResolvedSequenceParam] = field(default_factory=list)
    steps: list[str] = field(default_factory=list)


@dataclass
class ResolvedManifest:
    identity: HolonIdentity = field(default_factory=HolonIdentity)
    source_path: str = ""
    description: str = ""
    kind: str = ""
    build_runner: str = ""
    build_main: str = ""
    artifact_binary: str = ""
    artifact_primary: str = ""
    required_files: list[str] = field(default_factory=list)
    member_paths: list[str] = field(default_factory=list)
    skills: list[ResolvedSkill] = field(default_factory=list)
    sequences: list[ResolvedSequence] = field(default_factory=list)


def parse_holon(path: str | Path) -> HolonIdentity:
    """Parse a holon.proto file and return its identity."""
    return resolve_proto_file(path).identity


def parse_manifest(path: str | Path) -> ResolvedManifest:
    """Parse manifest fields from a specific holon.proto file."""
    return resolve_proto_file(path)


def resolve(root: str | Path) -> ResolvedManifest:
    """Resolve a holon's manifest from a directory or proto path."""
    candidate = Path(root).expanduser()
    if candidate.is_file():
        return resolve_proto_file(candidate)

    proto_files = _collect_proto_files(candidate)
    if not proto_files:
        raise FileNotFoundError(f"no {PROTO_MANIFEST_FILE_NAME} found in {candidate}")

    last_error: Exception | None = None
    for proto_path in proto_files:
        try:
            return resolve_proto_file(proto_path)
        except Exception as exc:  # pragma: no cover - guarded by tests via success path
            last_error = exc

    if last_error is not None:
        raise last_error
    raise FileNotFoundError(f"no {PROTO_MANIFEST_FILE_NAME} found in {candidate}")


def resolve_manifest(root: str | Path) -> tuple[HolonIdentity, str]:
    """Preserve the original identity-plus-source-path API."""
    resolved = resolve(root)
    return resolved.identity, resolved.source_path


def resolve_proto_file(path: str | Path) -> ResolvedManifest:
    """Resolve a specific holon.proto file into manifest data."""
    path = Path(path).expanduser().resolve()
    if path.name != PROTO_MANIFEST_FILE_NAME:
        raise ValueError(f"{path} is not a {PROTO_MANIFEST_FILE_NAME} file")

    text = path.read_text(encoding="utf-8")
    manifest_block = _extract_manifest_block(text)
    if manifest_block is None:
        raise ValueError(f"{path}: missing holons.v1.manifest option in holon.proto")

    manifest = ResolvedManifest(source_path=str(path))

    identity_block = _extract_block("identity", manifest_block)
    if identity_block is not None:
        manifest.identity.uuid = _scalar("uuid", identity_block)
        manifest.identity.given_name = _scalar("given_name", identity_block)
        manifest.identity.family_name = _scalar("family_name", identity_block)
        manifest.identity.motto = _scalar("motto", identity_block)
        manifest.identity.composer = _scalar("composer", identity_block)
        manifest.identity.clade = _scalar("clade", identity_block)
        manifest.identity.status = _scalar("status", identity_block)
        manifest.identity.born = _scalar("born", identity_block)
        manifest.identity.version = _scalar("version", identity_block)
        manifest.identity.proto_status = _scalar("proto_status", identity_block)
        manifest.identity.aliases = _compact_strings(_string_list("aliases", identity_block))

    lineage_block = _extract_block("lineage", manifest_block)
    if lineage_block is not None:
        manifest.identity.parents = _compact_strings(_string_list("parents", lineage_block))
        manifest.identity.reproduction = _scalar("reproduction", lineage_block)
        manifest.identity.generated_by = _scalar("generated_by", lineage_block)

    manifest.description = _scalar("description", manifest_block)
    manifest.identity.lang = _scalar("lang", manifest_block)
    manifest.kind = _scalar("kind", manifest_block)

    build_block = _extract_block("build", manifest_block)
    if build_block is not None:
        manifest.build_runner = _scalar("runner", build_block)
        manifest.build_main = _scalar("main", build_block)
        manifest.member_paths = _compact_strings(
            [
                _scalar("path", member_block)
                for member_block in _extract_repeated_blocks("members", build_block)
            ]
        )

    requires_block = _extract_block("requires", manifest_block)
    if requires_block is not None:
        manifest.required_files = _compact_strings(_string_list("files", requires_block))

    artifacts_block = _extract_block("artifacts", manifest_block)
    if artifacts_block is not None:
        manifest.artifact_binary = _scalar("binary", artifacts_block)
        manifest.artifact_primary = _scalar("primary", artifacts_block)

    manifest.skills = [
        ResolvedSkill(
            name=_scalar("name", skill_block),
            description=_scalar("description", skill_block),
            when=_scalar("when", skill_block),
            steps=_trimmed_strings(_string_list("steps", skill_block)),
        )
        for skill_block in _extract_repeated_blocks("skills", manifest_block)
    ]

    manifest.sequences = [
        ResolvedSequence(
            name=_scalar("name", sequence_block),
            description=_scalar("description", sequence_block),
            params=[
                ResolvedSequenceParam(
                    name=_scalar("name", param_block),
                    description=_scalar("description", param_block),
                    required=_bool("required", param_block),
                    default=_scalar("default", param_block),
                )
                for param_block in _extract_repeated_blocks("params", sequence_block)
            ],
            steps=_trimmed_strings(_string_list("steps", sequence_block)),
        )
        for sequence_block in _extract_repeated_blocks("sequences", manifest_block)
    ]

    return manifest


def find_holon_proto(root: str | Path) -> Path | None:
    """Locate a holon.proto file from a root directory or file path."""
    root = Path(root).expanduser()
    if root.is_file():
        return root.resolve() if root.name == PROTO_MANIFEST_FILE_NAME else None
    if not root.exists() or not root.is_dir():
        return None

    direct = root / PROTO_MANIFEST_FILE_NAME
    if direct.exists():
        return direct.resolve()

    api_v1 = root / "api" / "v1" / PROTO_MANIFEST_FILE_NAME
    if api_v1.exists():
        return api_v1.resolve()

    candidates = _collect_proto_files(root)
    return candidates[0] if candidates else None


def resolve_manifest_path(root: str | Path) -> Path:
    """Locate the nearest holon.proto relative to a proto directory or holon root."""
    root = Path(root).expanduser()
    search_roots = [root]
    if root.name == "protos":
        search_roots.append(root.parent)
    elif root.parent not in search_roots:
        search_roots.append(root.parent)

    for candidate_root in search_roots:
        candidate = find_holon_proto(candidate_root)
        if candidate is not None:
            return candidate

    raise FileNotFoundError(f"no {PROTO_MANIFEST_FILE_NAME} found near {root}")


def _collect_proto_files(root: Path) -> list[Path]:
    root = root.expanduser()
    if not root.exists() or not root.is_dir():
        return []

    files: list[Path] = []
    for path in sorted(root.rglob(PROTO_MANIFEST_FILE_NAME)):
        rel_parts = path.relative_to(root).parts
        if any(_should_skip_part(part) for part in rel_parts[:-1]):
            continue
        files.append(path.resolve())
    return files


def _should_skip_part(name: str) -> bool:
    return name in {".git", ".op", "node_modules", "vendor", "build", "gen", "testdata"} or (
        name.startswith(".") and not name.endswith(".holon")
    )


def _extract_manifest_block(source: str) -> str | None:
    match = _MANIFEST_OPTION_RE.search(source)
    if match is None:
        return None
    brace_index = source.find("{", match.start())
    if brace_index < 0:
        return None
    return _balanced_block_contents(source, brace_index)


def _extract_block(name: str, source: str) -> str | None:
    match = re.search(rf"\b{re.escape(name)}\s*:\s*\{{", source, re.DOTALL)
    if match is None:
        return None
    brace_index = source.find("{", match.start())
    if brace_index < 0:
        return None
    return _balanced_block_contents(source, brace_index)


def _extract_repeated_blocks(name: str, source: str) -> list[str]:
    blocks: list[str] = []
    for match in re.finditer(rf"\b{re.escape(name)}\s*:\s*\{{", source, re.DOTALL):
        brace_index = source.find("{", match.start())
        if brace_index < 0:
            continue
        block = _balanced_block_contents(source, brace_index)
        if block is not None:
            blocks.append(block)
    return blocks


def _scalar(name: str, source: str) -> str:
    quoted = re.search(rf'\b{re.escape(name)}\s*:\s*"((?:[^"\\]|\\.)*)"', source, re.DOTALL)
    if quoted is not None:
        return _unescape_proto_string(quoted.group(1))

    bare = re.search(rf"\b{re.escape(name)}\s*:\s*([^\s,\]}}]+)", source, re.DOTALL)
    if bare is not None:
        return bare.group(1)
    return ""


def _bool(name: str, source: str) -> bool:
    return _scalar(name, source).strip().lower() == "true"


def _string_list(name: str, source: str) -> list[str]:
    match = re.search(rf"\b{re.escape(name)}\s*:\s*\[(.*?)\]", source, re.DOTALL)
    if match is None:
        return []

    body = match.group(1)
    values: list[str] = []
    for token in re.finditer(r'"((?:[^"\\]|\\.)*)"|([^\s,\]]+)', body):
        quoted, bare = token.groups()
        if quoted is not None:
            values.append(_unescape_proto_string(quoted))
        elif bare is not None:
            values.append(bare)
    return values


def _balanced_block_contents(source: str, opening_brace: int) -> str | None:
    depth = 0
    inside_string = False
    escaped = False
    content_start = opening_brace + 1

    for index in range(opening_brace, len(source)):
        char = source[index]
        if inside_string:
            if escaped:
                escaped = False
            elif char == "\\":
                escaped = True
            elif char == '"':
                inside_string = False
            continue

        if char == '"':
            inside_string = True
        elif char == "{":
            depth += 1
        elif char == "}":
            depth -= 1
            if depth == 0:
                return source[content_start:index]

    return None


def _compact_strings(values: list[str]) -> list[str]:
    out: list[str] = []
    seen: set[str] = set()
    for value in values:
        trimmed = value.strip()
        if not trimmed or trimmed in seen:
            continue
        seen.add(trimmed)
        out.append(trimmed)
    return out


def _trimmed_strings(values: list[str]) -> list[str]:
    return [value.strip() for value in values if value.strip()]


def _unescape_proto_string(value: str) -> str:
    return value.replace(r"\"", '"').replace(r"\\", "\\")
